Before fix:
```python
$ pytest numpy/testing/tests/test_utils.py
================================================= test session starts ==================================================
platform linux -- Python 3.9.13, pytest-7.1.2, pluggy-1.0.0
PySide2 5.15.5 -- Qt runtime 5.15.4 -- Qt compiled 5.15.4
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
rootdir: /home/mark/git/numpy, configfile: pytest.ini
plugins: env-0.6.2, forked-1.4.0, hypothesis-6.47.1, qt-4.1.0, random-order-1.0.4, xdist-2.5.0, anyio-3.6.1, timeout-2.1.0
collected 132 items

numpy/testing/tests/test_utils.py .............................................................................. [ 59%]
......sss...........F.................................                                                           [100%]

======================================================= FAILURES =======================================================
____________________________________ TestAssertAllclose.test_error_message_unsigned ____________________________________

self = <numpy.testing.tests.test_utils.TestAssertAllclose object at 0x7f5fa557c310>

    def test_error_message_unsigned(self):
        """Check the message is formatted correctly for the decimal value.
           Also check the message when input includes inf or nan (gh12200)"""
        x = np.asarray([0, 1, 2], dtype='uint8')
        y = np.asarray([4, 4, 4], dtype='uint8')
        # Test with a different amount of decimal digits
        with pytest.raises(AssertionError) as exc_info:
            assert_allclose(x, y, atol=2)
        msgs = str(exc_info.value).split('\n')
>       assert_equal(msgs[4], 'Max absolute difference: 4')
E       AssertionError:
E       Items are not equal:
E        ACTUAL: 'Max absolute difference: 254'
E        DESIRED: 'Max absolute difference: 4'

exc_info   = <ExceptionInfo AssertionError('\nNot equal to tolerance rtol=1e-07, atol=2\n\nMismatched elements: 2 / 3 (66.7%)\nMax ...ce: 254\nMax relative difference: 63.5\n x: array([0, 1, 2], dtype=uint8)\n y: array([4, 4, 4], dtype=uint8)') tblen=3>
msgs       = ['', 'Not equal to tolerance rtol=1e-07, atol=2', '', 'Mismatched elements: 2 / 3 (66.7%)', 'Max absolute difference: 254', 'Max relative difference: 63.5', ...]
self       = <numpy.testing.tests.test_utils.TestAssertAllclose object at 0x7f5fa557c310>
x          = array([0, 1, 2], dtype=uint8)
y          = array([4, 4, 4], dtype=uint8)

numpy/testing/tests/test_utils.py:928: AssertionError
=============================================== short test summary info ================================================
FAILED numpy/testing/tests/test_utils.py::TestAssertAllclose::test_error_message_unsigned - AssertionError:
======================================= 1 failed, 128 passed, 3 skipped in 0.44s =======================================

```

--- a/numpy/testing/_private/utils.py
+++ b/numpy/testing/_private/utils.py
@@ -814,6 +814,9 @@ def assert_array_compare(comparison, x, y, err_msg='', verbose=True, header='',
                 # ignore errors for non-numeric types
                 with contextlib.suppress(TypeError):
                     error = abs(x - y)
+                    if np.issubdtype(x.dtype, np.unsignedinteger):
+                        error2 = abs(y - x)
+                        np.minimum(error, error2, out=error)
                     max_abs_error = max(error)
                     if getattr(error, 'dtype', object_) == object_:
                         remarks.append('Max absolute difference: '

--- a/numpy/testing/tests/test_utils.py
+++ b/numpy/testing/tests/test_utils.py
@@ -916,6 +916,20 @@ class TestAssertAllclose:
         a = np.array([[1, 2, 3, "NaT"]], dtype="m8[ns]")
         assert_allclose(a, a)
 
+    def test_error_message_unsigned(self):
+        """Check the the message is formatted correctly when overflow can occur
+           (gh21768)"""
+        # Ensure to test for potential overflow in the case of:
+        #        x - y
+        # and
+        #        y - x
+        x = np.asarray([0, 1, 8], dtype='uint8')
+        y = np.asarray([4, 4, 4], dtype='uint8')
+        with pytest.raises(AssertionError) as exc_info:
+            assert_allclose(x, y, atol=3)
+        msgs = str(exc_info.value).split('\n')
+        assert_equal(msgs[4], 'Max absolute difference: 4')
+
 
 class TestArrayAlmostEqualNulp:
 

